ion-schema/src/isl/isl_type_reference.rs

```diff
@@ -311,3 +326,90 @@ impl IslTypeRefImpl {
         }
     }
 }
+
+/// Represents a [variably occurring type reference] that will be used by `ordered_elements` and `fields` constraints.
+///  
+/// ```ion
+/// Grammar: <VARIABLY_OCCURRING_TYPE_ARGUMENT> ::= { <OCCURS>, <CONSTRAINT>... }
+///                                      | <TYPE_ARGUMENT>
+///
+///         <OCCURS> ::= occurs: <INT>
+///            | occurs: <RANGE_INT>
+///            | occurs: optional
+///            | occurs: required
+/// ```
+///
+/// [variably occurring type reference]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#variably-occurring-type-arguments
+#[derive(Debug, Clone, PartialEq)]
+pub struct IslVariablyOccurringTypeRef {
+    type_ref: IslTypeRefImpl,
+    occurs: Option<Range>,
```

Comment:
A variably occurring type reference always has an occurs range, though sometimes it is implicit. What do you think about modeling this not optional? 

To keep it nice to construct, we could have constructors such as:
```rust 
    pub fn optional(type_ref: IslTypeRef) -> Self {
        Self {
            type_ref: type_ref.type_reference,
            // TODO: add occurs optional here
        }
    }

    pub fn required(type_ref: IslTypeRef) -> Self {
        Self {
            type_ref: type_ref.type_reference,
            // TODO: add occurs required here
        }
    }
```